src/core/DY.Core.Trace.Base.fst

```diff
@@ -0,0 +1,349 @@
+   SMTPat (prefix tr2 i);
+   SMTPat (tr1 <$ tr2)]
+let rec prefix_prefix_eq tr1 tr2 i =
+  reveal_opaque (`%grows) (grows #label);
+  norm_spec [zeta; delta_only [`%prefix]] (prefix #label);
+  if length tr1 = length tr2 then ()
+  else (
+    let Snoc tr2_init _ = tr2 in
+    prefix_prefix_eq tr1 tr2_init i
+  )
+
+(*** Event in the trace predicates ***)
+
+/// Retrieve the event at some timestamp in the trace.
+
+val get_event_at:
+  #label_t:Type ->
+  tr:trace_ label_t -> i:timestamp{i < length tr} ->
+  trace_event_ label_t
+let rec get_event_at tr i =
+  if i+1 = length tr then
+    let Snoc _ last = tr in
+    last
+  else (
+    let Snoc tr_init _ = tr in
+    get_event_at tr_init i
+  )
+
+/// Has some particular event been triggered at a some particular timestamp in the trace?
+
+val event_at:
+  #label_t:Type ->
+  trace_ label_t -> timestamp -> trace_event_ label_t ->
+  prop
+let event_at tr i e =
+  if i >= length tr then
+    False
+  else
+    e == get_event_at tr i
+
+/// Has some particular event been triggered in the trace (at any timestamp)?
+
+val event_exists:
+  #label_t:Type ->
+  trace_ label_t -> trace_event_ label_t ->
+  prop
+let event_exists tr e =
+  exists i. event_at tr i e
+
+/// An event in the trace stays here when the trace grows.
+
+val event_at_grows:
+  #label_t:Type ->
+  tr1:trace_ label_t -> tr2:trace_ label_t ->
+  i:timestamp -> e:trace_event_ label_t ->
+  Lemma
+  (requires event_at tr1 i e /\ tr1 <$ tr2)
+  (ensures event_at tr2 i e)
+  [SMTPat (event_at tr1 i e); SMTPat (tr1 <$ tr2)]
+let rec event_at_grows #label_t tr1 tr2 i e =
+  reveal_opaque (`%grows) (grows #label_t);
+  norm_spec [zeta; delta_only [`%prefix]] (prefix #label_t);
+  if i >= length tr1 then ()
+  else if length tr1 >= length tr2 then ()
+  else (
+    let Snoc tr2_init _ = tr2 in
+    event_at_grows tr1 tr2_init i e
+  )
+
+/// Shorthand predicates.
+
+/// Has a message been sent on the network?
+
+val msg_sent_on_network: trace -> bytes -> prop
+let msg_sent_on_network tr msg =
+  event_exists tr (MsgSent msg)
+
+/// Has some state been stored by a principal?
+
+val state_was_set: trace -> principal -> state_id -> bytes -> prop
+let state_was_set tr prin sess_id content =
+  event_exists tr (SetState prin sess_id content)
+
+/// Has a principal been corrupt?
+
+val was_corrupt: #label_t:Type -> trace_ label_t -> principal -> state_id -> prop
+let was_corrupt tr prin sess_id =
+  event_exists tr (Corrupt prin sess_id)
+
+/// Has a (custom, protocol-specific) event been triggered at some timestamp?
+
+val event_triggered_at: trace -> timestamp -> principal -> string -> bytes -> prop
+let event_triggered_at tr i prin tag content =
+  event_at tr i (Event prin tag content)
+
+/// Has a (custom, protocol-specific) event been triggered (at any timestamp)?
+
+val event_triggered: trace -> principal -> string -> bytes -> prop
+let event_triggered tr prin tag content =
+  exists i. event_triggered_at tr i prin tag content
```

Comment:
Is there a reason that this doesn't use `event_exists` like its counterparts?

Reply:
Yes, it is to trigger this SMT pattern https://github.com/REPROSEC/dolev-yao-star-extrinsic/blob/b554c573480b0907b6c6309e6073042e71727335/src/core/DY.Core.Trace.Invariant.fst#L195-L206
If we used `event_exists` then the pattern `event_triggered_at` wouldn't show up and we wouldn't know that the event predicate was true on a prefix of the trace.